### ISSUE TITLE
IntrospectionError on Class/KClass recursive comparison

### ISSUE THREAD

**Issue by mkapiczy** [assertj/assertj#3289]
Hi

I encountered an issue with recursive comparison of Java/Kotlin class types. Basically those two tests are failing with the following error:

```
    @Test
    fun test() {
        val a: Any = Integer::class
        val b: Any = Long::class

        assertThat(a).usingRecursiveComparison().isEqualTo(b)
    }

    @Test
    fun test2() {
        val a: Any = Integer::class.java
        val b: Any = Long::class.java

        assertThat(a).usingRecursiveComparison().isEqualTo(b)
    }
```

```
org.assertj.core.util.introspection.IntrospectionError: 
Can't find any field or property with name 'timestamp'.
Error when introspecting properties was :
- No getter for property 'timestamp' in java.lang.ref.SoftReference 
Error when introspecting fields was :
- Unable to obtain the value of the field <'timestamp'> from <java.lang.ref.SoftReference@60b33bc>
	at app//org.assertj.core.util.introspection.PropertyOrFieldSupport.getSimpleValue(PropertyOrFieldSupport.java:88)
	at app//org.assertj.core.api.recursive.comparison.RecursiveComparisonConfiguration.dualValueForField(RecursiveComparisonConfiguration.java:631)
	at app//org.assertj.core.api.recursive.comparison.RecursiveComparisonConfiguration.lambda$getNonIgnoredActualFieldNames$3(RecursiveComparisonConfiguration.java:609)
	at java.base@17.0.8.1/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base@17.0.8.1/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base@17.0.8.1/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1707)
	at java.base@17.0.8.1/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base@17.0.8.1/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@17.0.8.1/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base@17.0.8.1/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@17.0.8.1/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at app//org.assertj.core.api.recursive.comparison.RecursiveComparisonConfiguration.getNonIgnoredActualFieldNames(RecursiveComparisonConfiguration.java:615)
	at app//org.assertj.core.api.recursive.comparison.RecursiveComparisonDifferenceCalculator.determineDifferences(RecursiveComparisonDifferenceCalculator.java:276)
	at app//org.assertj.core.api.recursive.comparison.RecursiveComparisonDifferenceCalculator.determineDifferences(RecursiveComparisonDifferenceCalculator.java:189)
	at app//org.assertj.core.api.RecursiveComparisonAssert.determineDifferencesWith(RecursiveComparisonAssert.java:1258)
	at app//org.assertj.core.api.RecursiveComparisonAssert.isEqualTo(RecursiveComparisonAssert.java:160)
```

Those are simple examples for reproduction of the issue I encountered it in my code while comparing more complex objects of a class with `KClass` property.

**assertj version:** 3.20.2

**Comment by quaff** [comment]
https://github.com/assertj/assertj/blob/5faf53f07ce3fe685e9d3da271f19eb52e829b7a/assertj-core/src/main/java/org/assertj/core/api/AbstractAssert.java#L1090

`AbstractAssert::usingRecursiveComparison` is not public API.

**Comment by scordio** [comment]
At first glance, I don't see a correlation with the error above.

Could you please elaborate, @quaff?

**Comment by quaff** [comment]
> At first glance, I don't see a correlation with the error above.
> 
> Could you please elaborate, @quaff?

It's not exposed API, `assertThat(a).usingRecursiveComparison().isEqualTo(b)` can not be compiled by javac, I think it's not expected behavior if it could be compiled by kotlin.

**Comment by mkapiczy** [comment]
It is public in `AbstractObjectAssert`

https://github.com/assertj/assertj/blob/5faf53f07ce3fe685e9d3da271f19eb52e829b7a/assertj-core/src/main/java/org/assertj/core/api/AbstractObjectAssert.java#L1219

**Comment by scordio** [comment]
Your Java example is unrelated to the Kotlin one from @mkapiczy since you're passing a `Class` instance to `assertThat()`, which implicitly exposes `Class` assertions where the recursive comparison is not offered by design.

**Comment by scordio** [comment]
> It seems that kotlin KClass is using AbstractObjectAssert and java class is using ClassAssert where it's protected.

Indeed: any type that doesn't have AssertJ specific assertions implicitly falls back to object assertions, where the recursive comparison is available.

> However, my original issue was not with a direct comparison of `KClass` types but more complex domain types with a nested `KClass` property field which is too complex to share in a simple reproducible example.

Correct, that's the use case we should analyze to see how we can make it better.

I'll mark this and the previous comment as off-topics to avoid stealing the focus from the original problem.

**Comment by scordio** [comment]
I executed the second example on the latest code:

```kotlin
@Test
fun test2() {
    val a: Any = Integer::class.java
    val b: Any = Long::class.java

    assertThat(a).usingRecursiveComparison().isEqualTo(b)
}
```

which fails with:

```
java.lang.AssertionError: 
Expecting actual:
  java.lang.Integer
to be equal to:
  long
when recursively comparing field by field, but found the following difference:

Top level actual and expected objects differ:
- actual value  : java.lang.Integer
- expected value: long
```

almost the same way as what happens in Java:

```java
@Test
void test2() {
  Object a = Integer.class;
  Object b = Long.class;

  assertThat(a).usingRecursiveComparison().isEqualTo(b);
}
```

```
java.lang.AssertionError: 
Expecting actual:
  java.lang.Integer
to be equal to:
  java.lang.Long
when recursively comparing field by field, but found the following difference:

Top level actual and expected objects differ:
- actual value  : java.lang.Integer
- expected value: java.lang.Long
```

It seems that the Kotlin `b` gets the `Class` instance of the `long` primitive type instead of the corresponding wrapper type. I'm not familiar enough with Kotlin to explain why the wrapper type instance is assigned to `a` while the primitive type instance is assigned to `b`, but I consider it unrelated to AssertJ and so what I experience looks good to me.

**Comment by scordio** [comment]
Running:

```kotlin
@Test
fun test() {
    val a: Any = Integer::class
    val b: Any = Long::class

    assertThat(a).usingRecursiveComparison().isEqualTo(b)
}
```

or:

```kotlin
@Test
fun test() {
    val a = Integer::class // KClass<Integer>
    val b = Long::class    // KClass<Long>

    assertThat(a).usingRecursiveComparison().isEqualTo(b)
}
```

yields:

```
java.lang.AssertionError: 
Expecting actual:
  class java.lang.Integer (Kotlin reflection is not available)
to be equal to:
  long (Kotlin reflection is not available)
when recursively comparing field by field, but found the following difference:

field/property 'jClass' differ:
- actual value  : java.lang.Integer
- expected value: long
```

Aside from the `long` vs. `java.lang.Long` type also mentioned in the previous comment, the behavior looks good to me.

**Comment by scordio** [comment]
In 7406133f21ed66e3908a50c581ad0f0aaf059aa6, we decided to avoid the recursive comparison for JDK types and just rely on what `equals` does, even if some JDK classes don't override it. See also #2466 and #2450 for more background.

This change was released in 3.23.0.

@mkapiczy please upgrade to 3.23.0 or 3.24.2 (latest) to get the behavior I mentioned above.
